### ISSUE TITLE
Add ability to manually provide pattern lineage info

### ISSUE THREAD

**Issue by james-nash** [pattern-lab/patternlab-node#925]
### Summary
Sometimes it is not practical to compose larger patterns (e.g. molecules or organisms) from smaller ones via Mustache includes (or their equivalent, if using another template engine). In such situations, it's often easier to simply copy-paste the markup of the smaller pattern into the larger one.

Unfortunately, doing so means that components that Pattern Lab is unable to detect the lineage since there are no explicit template includes and you will therefore not get any links between those patterns in the pattern info panel. That's a pity, because those links in the generated pattern library can be really helpful to understand how and where components are being used.

I am therefore proposing that there should be a means to manually specify lineage between patterns.

### Example
Consider things like tables. Many pattern libraries will have a table pattern to showcase whatever styling it has. So, image we make a simple `table.mustache` like this:

```mustache
<table>
{{ #rows }}
  <tr>
    {{ #cols }}
      <td>{{ content }}</td>
    {{ /cols }}
  </tr>
{{ /rows }}
</table>
```

There are a few issues with this:

* What happens if we want to include `<th>` elements for header rows or columns? We could extend our pattern, but it would quickly get more complex and the corresponding data in `data.json` would also begin to get unwieldy
* We've limited the content of cells to plain text. What happens if they need to include other mark-up or even other components?
* If a larger component wants to include more than one table content, but give each instance different content, that becomes very tricky.

In practise what I've done and seen others do in this kind of scenario is accept that it's not worth the effort of making some über-complex `table.mustache` that other patterns can include and tailor to whatever they may need. Instead, you make a dumb `table.mustache` with some hard-coded content simply to show what table styles look like. Then, any other components that make contain tables just have the necessary markup written directly into their Mustache files.

(I've used tables as an example that people can hopefully relate to. However, I've faced this "to include or not to include" conundrum with all kinds of things)

Unfortunately, doing so means that components that contain tables don't actually have an include to the `table.mustache`. Consequently PL can't auto-detect the relationship between those patterns and you don't get those lovely "Pattern X include Pattern Y" links.

### Possible approach
We can already provide some meta data about patterns via the frontmatter in a pattern's markdown file - for example `state` and `title`. Perhaps this could be extended, and PL could add support for an `includes` property that takes the name of a pattern (or an array of pattern names) that PL can then treat as though they were actually being included via the template. For example:

`product-comparison-table.md`:
```md
---
title: Product comparison table
includes: atoms-table
---
Lorem ipsum...
```

Pattern Lab could then augment the lineage info it figures out automatically via proper template includes with this manually provided information. The `product-comparison-table`'s pattern info panel would then contain a link to `atoms-table` along with any other patterns it includes. Likewise, `atoms-table`'s pattern info would contain a link back to `product-comparison-table`.

What do you think? If this isn't too hard to implement and others would find it useful, then I'd love to see this in a future PL release!

**Comment by bmuenzenmeyer** [comment]
I like this idea and think it has legs, as long as others agree.

**Comment by bradfrost** [comment]
@james-nash This is fascinating; thanks for sharing! 

> Sometimes it is not practical to compose larger patterns (e.g. molecules or organisms) from smaller ones via Mustache includes (or their equivalent, if using another template engine). In such situations, it's often easier to simply copy-paste the markup of the smaller pattern into the larger one.

I'm not really sure I follow here. Creating smaller components and including them in larger components is the core of component driven development. If a pattern is manually copied and pasted in a larger component, it's not truly using that component. I guess I don't understand why one would copy and paste a pattern rather than including the smaller component. That doesn't keep things DRY (don't repeat yourself), which makes the lineage false since it's technically not using the reusable component. 

> What happens if we want to include <th> elements for header rows or columns? We could extend our pattern, but it would quickly get more complex and the corresponding data in data.json would also begin to get unwieldy

I don't really see a problem with adding complexity to a component that can be complex. Adding the following to the `table.mustache` pattern will add a `<thead>` to the table:

```
{{# th }}
<thead>
<tr>
{{# theadings }}
<th>{{ th }}</th>
{{# theadings }}
</tr>
</thead>
{{/ thead }}
```

Does that make the table component more complex? Yes it does. But I'm not sure I understand the issue with introducing that complexity to a component that can have that option. 

Also worth noting that if you wanted to chunk out the table header into its own component, you can create a `table-header.mustache` component and [hide it from Pattern Lab's navigation](https://patternlab.io/docs/pattern-hiding.html), like `_table-header.mustache`. 

> We've limited the content of cells to plain text. What happens if they need to include other mark-up or even other components?

Mustache unfortunately doesn't support container components, but other templating engines like Handlebars, React, Twig, and others do. This is a limitation of Mustache rather than a PL problem.

> If a larger component wants to include more than one table content, but give each instance different content, that becomes very tricky.

This can be handled by wrapping the include in a specific ID name. In our templates, we'll often include multiple instances of a component, and delineate them like so:

```
{{# productComparisonTable }}
{{> organisms-table }}
{{/ productComparisonTable }}

{{# anotherTable }}
{{> organisms-table }}
{{/ anotherTable }}
```

In order to document and express specific instances of a pattern in the pattern library, you can use [pseudo-patterns](https://patternlab.io/docs/pattern-pseudo-patterns.html). So you can have your `table.mustache` component that provides the source of truth for that component. From there, you can create instances like `table~product-comparison.json`, `table~with-header.json`, `table~zebra-stripes.json`, etc to articulate all the potential content/style/behavior variations you need. 

All of this is a long-winded way of saying that Pattern Lab has conventions in place for achieving a lot of this stuff. I'd be hesitant to introduce a manual lineage as it introduces a false sense of pattern inclusion. I very well could be misinterpreting all of this, so still open to hearing more about it.

**Comment by james-nash** [comment]
Hi @bradfrost. Thanks so much for taking the time to write that detailed reply!

While I completely agree that, ideally, it's better to make suitably flexible component templates and actually include them where they get used, I have come across instances where this doesn't get done (and am guilty of making them myself sometimes too). While my colleagues and I do strive for the "holy grail" of Pattern Lab templates being shared with actual production sites, this isn't always practical. There are a number of reasons, but ones I've come across are:

* Large organisations using a variety of tech stacks, so whichever template engine we use in PL, there will at least be instances of consumers using something else
* Inability to liaise with downstream dev teams (either because they're different companies, or simply because a project was set up in a way where they aren't around to talk to until after we've done our bit in PL and have moved on).

So, while I know it's suboptimal (and do try to avoid it wherever I can), we sometimes only use PL to _show_ how things should look, but rely on others to build what they see there in whatever way they see fit. Other times we _are_ sharing code with production, but it's only the CSS & JS, not the templates themselves.

In those scenarios, we sometimes end up sacrificing the flexibility of our pattern templates in PL for speed or convenience. So, in my table example, we might do what I described because we only use tables in a handful of places and it's just easier to hardcode them rather than making a more flexible, but complex, table pattern.

The motivation for my feature request is therefore as a work-around to still get some useful lineage links in those kinds of situations. (Because, for people viewing the pattern library, I think those links can be very helpful to highlight the (intended) relationships between patterns)

Long story short, I appreciate that there are better ways to do things in PL that avoid the need for manual lineage. However, I believe there are people out there that would benefit from this if it's not too much hassle to add (and maintain).

**Comment by bradfrost** [comment]
Thanks for the additional info, @james-nash. 

The situations you're spelling out are very much ones I've encountered. In my own work, we often use Pattern Lab as the [workshop](http://bradfrost.com/blog/post/the-workshop-and-the-storefront/) to create the canonical version of the design system. There can be [many other tech-specific flavors of the design system](http://bradfrost.com/blog/post/managing-technology-agnostic-design-systems/), but it's not Pattern Lab's job to manage them.

These situations are real, but it still seems like a workflow issue and/or something that documentation can help solve, rather than looking to Pattern Lab to solve these issues. I don't think that building a feature into Pattern Lab that creates an untrue lineage is the solution to this problem. I think perhaps [adding that information as documentation](https://patternlab.io/docs/pattern-documenting.html) can solve the problem. That documentation lives right alongside the lineage, so perhaps that can give you a way to address the problem?

**Comment by james-nash** [comment]
Hi @bradfrost. That's fair.

Thanks again to you and @bmuenzenmeyer for taking the time to consider and discuss my feature suggestion so thoroughly. Much appreciated!
